client/state/data-layer/wpcom/all-domains/index.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * External Dependencies
+ */
+import { translate } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import { convertToCamelCase } from 'state/data-layer/utils';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import { ALL_DOMAINS_REQUEST } from 'state/action-types';
+import { errorNotice } from 'state/notices/actions';
+import { http } from 'state/data-layer/wpcom-http/actions';
+import {
+	getAllDomainsRequestFailure,
+	getAllDomainsRequestSuccess,
+} from 'state/all-domains/actions';
+import { registerHandlers } from 'state/data-layer/handler-registry';
+
+export const getAllDomains = ( action ) => {
+	return http(
+		{
+			method: 'GET',
+			path: `/all-domains`,
+		},
+		action
+	);
+};
+
+export const getAllDomainsError = ( action, error ) => {
+	return [
+		errorNotice( translate( 'Failed to retrieve all domains' ) ),
+		getAllDomainsRequestFailure( error ),
+	];
+};
+
+export const getAllDomainsSuccess = ( action, response ) => {
+	if ( response ) {
+		return getAllDomainsRequestSuccess( response.domains );
+	}
+	return getAllDomainsError( action, 'Failed to retrieve your domains. No response was received' );
```

Comment:
I know it shouldn't be possible to end up in that situation but maybe we should not handle it as an error - not getting any domains might also mean that you have no domains. Technically you should not be able to access the "all my sites" page in that case but who knows what will happen... :D I'm fine though to fix this in a follow up PR

Reply:
In such a case the **_response_** from the endpoint should _return_ an explicit empty array.  That is clearer to wrangle. Not receiving a response could be as a result of a timeout.